tests/unix_listener.rs

```diff
@@ -0,0 +1,296 @@
+#![cfg(unix)]
+#[macro_use]
+mod util;
+
+use mio::net::UnixListener;
+use mio::{Interests, Poll, Token};
+use std::io::{self, Read};
+use std::os::unix::io::AsRawFd;
+use std::os::unix::net;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Barrier};
+use std::thread;
+use tempdir::TempDir;
+use util::{
+    assert_send, assert_sync, assert_would_block, expect_events, expect_no_events, init_with_poll,
+    ExpectEvent,
+};
+
+const DEFAULT_BUF_SIZE: usize = 64;
+const TOKEN_1: Token = Token(0);
+const TOKEN_2: Token = Token(1);
+
+#[test]
+fn unix_listener_send_and_sync() {
+    assert_send::<UnixListener>();
+    assert_sync::<UnixListener>();
+}
+
+#[test]
+fn unix_listener_smoke() {
+    #[allow(clippy::redundant_closure)]
+    smoke_test(|path| UnixListener::bind(path));
+}
+
+#[test]
+fn unix_listener_from_std() {
+    smoke_test(|path| {
+        let listener = assert_ok!(net::UnixListener::bind(path));
+        // `std::os::unix::net::UnixStream`s are blocking by default, so make sure
+        // it is in non-blocking mode before wrapping in a Mio equivalent.
+        assert_ok!(listener.set_nonblocking(true));
+        Ok(UnixListener::from_std(listener))
+    })
+}
+
+#[test]
+fn unix_listener_try_clone_same_poll() {
+    let (mut poll, mut events) = init_with_poll();
+    let barrier = Arc::new(Barrier::new(3));
+    let dir = assert_ok!(TempDir::new("unix_listener"));
+    let path = dir.path().join("any");
+
+    let listener1 = assert_ok!(UnixListener::bind(&path));
+    let listener2 = assert_ok!(listener1.try_clone());
+    assert_ne!(listener1.as_raw_fd(), listener2.as_raw_fd());
+
+    let handle_1 = open_connections(path.clone(), 1, barrier.clone());
+    assert_ok!(poll
+        .registry()
+        .register(&listener1, TOKEN_1, Interests::READABLE));
+    assert_ok!(poll
+        .registry()
+        .register(&listener2, TOKEN_2, Interests::READABLE));
+    expect_events(
+        &mut poll,
+        &mut events,
+        vec![
+            ExpectEvent::new(TOKEN_1, Interests::READABLE),
+            ExpectEvent::new(TOKEN_2, Interests::READABLE),
+        ],
+    );
+
+    assert_ok!(listener1.accept());
+
+    let handle_2 = open_connections(path.clone(), 1, barrier.clone());
+    expect_events(
+        &mut poll,
+        &mut events,
+        vec![
+            ExpectEvent::new(TOKEN_1, Interests::READABLE),
+            ExpectEvent::new(TOKEN_2, Interests::READABLE),
+        ],
+    );
+
+    assert_ok!(listener2.accept());
+
+    assert_would_block(listener1.accept());
+    assert_would_block(listener2.accept());
+
+    assert!(assert_ok!(listener1.take_error()).is_none());
+    assert!(assert_ok!(listener2.take_error()).is_none());
```

Comment:
Left a [comment](https://github.com/tokio-rs/mio/pull/1145) about the value I see in `assert_ok!` (and `assert_err!`)

Reply:
I'd like to move conversation about something like this to that PR though. I'd rather get this merged so that UDS is properly tested in master as soon as possible.